[Diagnostics]: Diagnostics.AspNetCore3 for supporting ASP.NET Core 3.1 (#5399)
googleapis/google-cloud-dotnet: apis/Google.Cloud.Diagnostics.AspNetCore3/docs/index.md

{{title}}

> `Google.Cloud.Diagnostics.AspNetCore` is an ASP.NET Core instrumentation library for Google Stackdriver.

Reviewer: ```suggestion
`Google.Cloud.Diagnostics.AspNetCore3` is an ASP.NET Core instrumentation library for Google Stackdriver.
```

It allows for simple integration of Stackdriver into ASP.NET applications with minimal code changes.

This package supports Stackdriver Error Reporting, Stackdriver
Logging and Stackdriver Trace, and integrates with ASP.NET Core 2.1

> (which is the long-term support release of ASP.NET Core 2.x) and above.

Reviewer: ```suggestion
Logging and Stackdriver Trace, and integrates with ASP.NET Core 3.1
(which is the long-term support release of ASP.NET Core 3.x) and above.
```

{{version}}

{{installation}}

{{auth}}

See [API Permissions for `entries.write`](https://cloud.google.com/logging/docs/access-control#api-permissions)
for the permissions needed for Logging and Error Reporting.

See [API Permissions for PatchTraces](https://cloud.google.com/trace/docs/iam#api_permissions)
for the permissions needed for Tracing.

# Note

> The Google.Cloud.Diagnostics.AspNetCore package attempts to collect the filename and line number when

Reviewer: ```suggestion
The Google.Cloud.Diagnostics.AspNetCore3 package attempts to collect the filename and line number when
```

entries are collected. However, to be able to collect this information PDBs must be included with
the deployed code.

# Note
When running on environments that limit or disable CPU usage for background activities, for instance
[Google Cloud Run](https://cloud.google.com/run/docs/tips/general#avoiding_background_activities), take care
not to use the timed buffer options for any of Logging, Tracing or Error Reporting. Take into account
that the timed buffer is used for all of these components by default so you will need to explicitly

> configure the buffers by using the `Google.Cloud.Diagnostics.AspNetCore.LoggerOptions`,

Reviewer: ```suggestion
configure the buffers by using the `Google.Cloud.Diagnostics.AspNetCore3.LoggerOptions`,
```

`Google.Cloud.Diagnostics.Common.TraceOptions` and `Google.Cloud.Diagnostics.Common.ErrorReportingOptions` classes.
Below you'll find examples of how to configure the buffers.

# Getting started

## Initializing Google Diagnostics

The easiest way to initialize Google Diagnostics services is using

> the `UseGoogleDiagnostics` extentension method on `IWebHostBuilder`.

Reviewer: Is this still IWebHostBuilder? I lose track of what's deprecated...

This configures Logging, Tracing and Error Reporting middleware.

If your application is runnng on Google App Engine, Google
Kubernetes Engine, Google Cloud Run or Google Compute Engine, you

> don't need to provide a value for `ProjectID`, `Service` and

Reviewer: Is it ProjectID or ProjectId? (I'd expect the latter.) If this needs correcting, search-and-replace it - I haven't looked for all occurrences.

activated as a one off, if you are having trouble trying to find
your logs in the GCP Console, and not as a permanent feature in
production code. To deactivate this feature simply stop passing a
`System.IO.TextWriter` as part of the options when creating a Google
Logger.

# Tracing

## Initializing Tracing

{{sample:Trace.RegisterGoogleTracer}}

## Troubleshooting Tracing

Just as with logging, trace is most easily diagnosed by removing
buffering and propagating exceptions immediately.

{{sample:Trace.Troubleshooting}}

The options can be specified wherever you are configuring tracing.

## Tracing in MVC Controllers

To use the `IManagedTracer` in MVC controllers you can either inject the singleton instance of 
`IManagedTracer` into the controller's constructor (see `TraceSamplesConstructorController`) or you
can in inject the `IManagedTracer` into the action method using the `[FromServices]` attribute
(see `TraceSamplesMethodController`).

{{sample:Trace.TraceMVCConstructor}}

{{sample:Trace.TraceMVCMethod}}

## Manual Tracing

{{sample:Trace.UseTracer}}

{{sample:Trace.UseTracerRunIn}}

## Trace Outgoing HTTP Requests (recommended)

> The [recommended way of creating HttpClient](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-2.1) in ASP.NET Core 2.0 and upwards is to use the

Reviewer: That links 
```suggestion
The [recommended way of creating HttpClient](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-3.1) in ASP.NET Core 2.0 and upwards is to use the
```

(Deliberately haven't changed the 2.0 and upwards.)

{{title}}

> `Google.Cloud.Diagnostics.AspNetCore3` is an ASP.NET Core instrumentation library for Google Logging, Error Reporting and Tracing.

Reviewer: Whoops, think I've left a bunch of comments before in an earlier commit...

Author: Yep, no worries, I went through those as well in case there was something there I'd missed.

> It allows for simple integration of Google observavility components into ASP.NET Core 3.1+ applications with minimal code changes.

Reviewer: ```suggestion
It allows for simple integration of Google observability components into ASP.NET Core 3.1+ applications with minimal code changes.
```

Author: Done

well as the `Service` and `Version` with which to identify your
application.

{{sample:Diagnostics.UseGoogleDiagnostics_Core3}}

You can still initialize the separate components using the extension
methods below. This can be useful if you only need to use some of

> the observavility components.

Reviewer: ```suggestion
the observability components.
```

Author: Done

## Initializing Logging

Logging can be initialized using `ConfigureServices` when

> constructing an `IHostBuilder`:

Reviewer: Same comments as above.

Author: Same, I've used:

```
When configuring an `IHostBuilder` Logging can be initialized in
two different ways:

Using `ConfigureServices`:

{{sample:Logging.RegisterGoogleLogger2_Core3}}

Or using `ConfigureLogging`:

{{sample:Logging.RegisterGoogleLogger3_Core3}}
```

## Trace Outgoing HTTP Requests (recommended)

> The [recommended way of creating HttpClient](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-3.1) in ASP.NET Core 3.1 and upwards is to use the

Reviewer: As before - I'd make the link 3.1, but keep it as 2.0 and upwards in the text.

Author: Yes, agreed.
